### DISCUSSION TITLE
Templating added on import?

### DISCUSSION THREAD

**Discussion by SubZane** [mockoon/mockoon#366] [Help] [answered]
It seems to be somewhat implemented. {{faker 'random.number'}} is added on numbers, but so many values are empty it would be really nice to get some dummy data on all of them. Adding it manually on every endpoints really takes alot of time.

Is this possible now, and I just missed it?

**Comment by 255kb** [comment]
Hi @SubZane, I converted the issue to a discussion as there is no bug or feature identified yet :) 

If I understand well, you are talking about OpenAPI import, which creates examples bodies?

**Reply by SubZane** [reply]
Yes. After importing my open api scheme I get the following. Some of the keys get faker-values, but some are just empty. I dunno if I missed something or if it's a limitation in the product?

```
[
  {
    "id": {{faker 'random.number'}},
    "uuid": "{{faker 'random.uuid'}}",
    "entity_name": "",
    "entity_type": "",
    "operation_type": "",
    "status": "{{oneOf (array 'P' 'N' 'R' 'C' 'E')}}",
    "create_date": "{{faker 'date.recent' 365}}",
    "start_date": "{{faker 'date.recent' 365}}",
    "end_date": "{{faker 'date.recent' 365}}",
    "log": "",
    "product": {{faker 'random.number'}},
    "user": {{faker 'random.number'}}
  }
]
```

**Reply by SubZane** [reply]
Also, since it's an array it would be nice if it was automatically wrapped in a repeat {{# repeat 10 }} {{/ repeat}}

**Reply by brentlyonsyapi** [reply]
I was hoping to add on to this, as this very scenario has become a requirement for our company (where we are happily working on integrating Mockoon!).. My hope was that there was a way to create some sort of simple "Global import rules", such as "if field = 'operation_type', then key = "{{oneOf (array 'P' 'N' 'R' 'C' 'E')}}", or "if endpoint returns a collection, then wrap results in a {repeat 10} loop". We are actually planning on integrating Mockoon with Jenkins, so that whenever we make an update to our APIs and we merge to master, an OpenAPI json spec file will be delivered to our Mockoon server. Any thoughts on a global-import-rule thing, where the user can define what goes into the value of the field?

Either way, thanks so much for creating Mockoon! At least for our requirements, it has proved to be the best free mock API generator out there.

**Reply by 255kb** [reply]
Hi @brentlyonsyapi thank you for your kind words :)
There is no such mechanism of global import templating rules right now. It's also not planned and would require, I guess, some sort of plugin system. Which is also not on the roadmap. 
I think there is no alternative yet, except maybe to add these templating snippets to the generated body using some kind of scripting.

**Comment by 255kb** [comment, marked as answer]
This is definitely not implemented yet. The thing is, we get from the OpenAPI schema some very limited info about the property types (string, date, etc.), and for each of them we replaced by a faker helper when it was obvious. However, behind `string` we can have many things: cities, firstname, etc. We would need to hardcode a list of property names with their corresponding faker helper, which is 100% doable but not that straightforward. For example, which helper would you use for "entity_name" or "entity_type"? I would say "company.companyName" for the first one, but I am not sure for the second one. 
Also, what happens with "lat", "long" that are numbers? We definitely have faker helpers for this, but we would need to read the type (`number`) and the property name (which could be "lat" or "latitude" or even something else), and we may still got "false positives". 
So it's not an easy task and it may be really time consuming. 

That being said, you are completely free to open an issue (please use the issue template), with more details. I can only say that this won't and cannot be on the short term roadmap. If you are curious about what we are currently working on you can check the roadmap there: https://github.com/orgs/mockoon/projects/2 
The CLI is the most requested feature right now and is getting all our attention. 
However, if you want to contribute and add some rules you can definitely work on it and submit a PR, we would be really happy to help you add this to Mockoon. 
Relevant code is there https://github.com/mockoon/mockoon/blob/aac1684d88120cb0dea3f718b21d8a4341a54836/src/app/services/openapi-converter.service.ts#L451 and it shouldn't be too hard to add more to the list depending on the field name and not only the field `format` or `type`.

**Reply by SubZane** [reply]
good points! Impossible to guess the values based on the key names without some mapping...hmm

**Reply by 255kb** [reply]
yes, I don't say it's impossible, but among all the things we have on our plate we cannot focus on this right now. And I say "we", because since 2 weeks we are lucky to be 2 people in the maintainers team, but it's still a side project for both of us. So we work during nights and weekends 😄
